json: Added element descriptor macro and C++ support progress

Sir-Branch opened the pull request:
Opening this PR to start the discussion, code formatting is not perfect.

 I'm working towards cleaning up the JSON library and making it more maintainable, on my high priorities is compatibility with C++.

What are the Zephyr guidelines towards having empty macro arguments? In this specific use case the union_ parameter in JSON_ELEMENT_DESCR may be empty, this is to insure a single point of use for ``` (struct json_obj_descr[])```

The underlying issue which motivates this PR is issue #18551, in this case macros that utilize ``` struct json_obj_descr[] ``` aren't able to run on C++. A possible patch is to replace all these occurrents for ``` const struct json_obj_descr[1] ```.

martinjaeger reviewed (commented):
If the new macros should not be used from the application, they should probably be prefixed with `Z_` so that they are excluded from the documentation (see fb4d9c04111b145e08fd6848d7a4d595705afafe).

Otherwise looks good to me.

Empty arguments seem to be valid C99, [see here](https://stackoverflow.com/questions/7666344/are-empty-macro-arguments-legal-in-c11). So I don't see any issue with that.

Sir-Branch commented:
I appreciate the insight, thanks for the help!

- I went ahead and prefixed with Z_
- Improved C++ support by changing from ` struct json_obj_descr[]` to `const struct json_obj_descr[1]`
- Rebased to pass CI

martinjaeger commented on include/data/json.h:
```diff
@@ -173,6 +173,54 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
+ * @param union_ Optional macro argument containing array or object descriptor
+ */
+#define Z_JSON_ELEMENT_DESCR(struct_, len_field_, elem_type_, union_) \
+	(const struct json_obj_descr[1]) \
```
Whether `const struct json_obj_descr[]` or `const struct json_obj_descr[1]` is better/valid is beyond my C vs. C++ knowledge. Maybe @pabigot could have a quick look, as you provided related PRs for Bluetooth?

pabigot replied:
Since when I fixed the C++ problem with using an address of temporary I used `type[]` I don't know of any reason why it should change to `type[1]`, but newer versions of GNUC might have a problem.  In context `[]` should just cause the compiler to infer the array length from the single-element value.

Sir-Branch replied:
Agreed, there really shouldn't be a difference between ```type[]``` and ```type[1]``` as compiler will infer the length. Specifying the length could cause maintenance issue down the road, I'll update to ```const struct json_obj_descr[]```. I really  appreciate the insight

Sir-Branch commented:
@martinjaeger rebased to avoid conflict due to #38061: Updated use of `JSON_TOK_LIST_` to `JSON_TOK_ARRAY_`

cfriedt commented on include/data/json.h:
```diff
@@ -177,6 +177,54 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 
+/**
+ * @brief Internal helper macro to declare an element descriptor
+ *
```
May be good to apply the `@internal` tag as well

cfriedt commented on include/data/json.h:
```diff
@@ -177,6 +177,54 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
+	}
+
+/**
+ * @brief Internal helper macro to declare an array descriptor
```
May be good to apply the `@internal` tag as well

cfriedt commented on include/data/json.h:
```diff
@@ -177,6 +177,54 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
+	}
+
+/**
+ * @brief Internal helper macro to declare an object descriptor
```
May be good to apply the `@internal` tag as well

Sir-Branch commented:
@cfriedt Thank you for the review have updated accordingly.

I really appreciate all the help you and @martinjaeger have provided, you've both made the process of contributing to Zephyr seamless.

Sir-Branch commented:
@mrfuchs - Would it be possible to get a review?

mrfuchs commented on include/data/json.h:
```diff
@@ -506,21 +518,9 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
-				} }, \
+				.element_descr =  Z_JSON_ELEMENT_DESCR(struct_, len_field_, \
+					JSON_TOK_OBJECT_START, \
+					JSON_DESCR_OBJECT(elem_descr_, elem_descr_len_)), \
```
```suggestion
					Z_JSON_DESCR_OBJ(elem_descr_, elem_descr_len_)), \
```

Sir-Branch replied:
Thank you, this one totally slipped!

mrfuchs reviewed (changes requested):
`JSON_OBJ_DESCR_OBJ_ARRAY_NAMED` needs to be fixed, otherwise LGTM.

mrfuchs commented on include/data/json.h:
```diff
@@ -506,21 +518,9 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
-						}, \
-					}, \
-				} }, \
+				.element_descr =  Z_JSON_ELEMENT_DESCR(struct_, len_field_, \
```
```suggestion
				.element_descr = Z_JSON_ELEMENT_DESCR(struct_, len_field_, \
```

Sir-Branch replied:
Nice catch, thank you!

Sir-Branch committed:
json: Added element descriptor macro and C++ support progress

Added element descriptor macro with internal helper macros for arrays
and objects.
Improved support for C++ by switching to const struct in element descr

Signed-off-by: Ramiro Merello <rmerello@itba.edu.ar>

Sir-Branch commented:
@mrfuchs thank you for the review. Have updated accordingly 😄

Sir-Branch commented:
@cfriedt do you think this can be merged in? Not sure why @mrfuchs approval isn't counting, even though he's a collaborator. Maybe he doesn't have write access? (Apologies if this is obvious, this is new to me)

cfriedt commented:
@mbolivar-nordic - I think that @mrfuchs is a collaborator for JSON. Any idea why have a approval might not work?
